lightning/src/routing/router.rs

```diff
@@ -2194,6 +2253,326 @@ mod tests {
+		let route = get_route(&our_id, &net_graph_msg_handler.network_graph.read().unwrap(), &nodes[6], None, None, &empty_last_hops(&nodes).iter().collect::<Vec<_>>(), 100, 42, Arc::clone(&logger)).unwrap();
+		assert_eq!(route.paths[0].len(), 5);
+
+		assert_eq!(route.paths[0][0].pubkey, nodes[1]);
+		assert_eq!(route.paths[0][0].short_channel_id, 2);
+		assert_eq!(route.paths[0][0].fee_msat, 100);
+		assert_eq!(route.paths[0][0].cltv_expiry_delta, (4 << 8) | 1);
+		assert_eq!(route.paths[0][0].node_features.le_flags(), &id_to_feature_flags(2));
+		assert_eq!(route.paths[0][0].channel_features.le_flags(), &id_to_feature_flags(2));
+
+		assert_eq!(route.paths[0][1].pubkey, nodes[2]);
+		assert_eq!(route.paths[0][1].short_channel_id, 4);
+		assert_eq!(route.paths[0][1].fee_msat, 0);
+		assert_eq!(route.paths[0][1].cltv_expiry_delta, (6 << 8) | 1);
+		assert_eq!(route.paths[0][1].node_features.le_flags(), &id_to_feature_flags(3));
+		assert_eq!(route.paths[0][1].channel_features.le_flags(), &id_to_feature_flags(4));
+
+		assert_eq!(route.paths[0][2].pubkey, nodes[4]);
+		assert_eq!(route.paths[0][2].short_channel_id, 6);
+		assert_eq!(route.paths[0][2].fee_msat, 0);
+		assert_eq!(route.paths[0][2].cltv_expiry_delta, (11 << 8) | 1);
+		assert_eq!(route.paths[0][2].node_features.le_flags(), &id_to_feature_flags(5));
+		assert_eq!(route.paths[0][2].channel_features.le_flags(), &id_to_feature_flags(6));
+
+		assert_eq!(route.paths[0][3].pubkey, nodes[3]);
+		assert_eq!(route.paths[0][3].short_channel_id, 11);
+		assert_eq!(route.paths[0][3].fee_msat, 0);
+		assert_eq!(route.paths[0][3].cltv_expiry_delta, (8 << 8) | 1);
+		// If we have a peer in the node map, we'll use their features here since we don't have
+		// a way of figuring out their features from the invoice:
+		assert_eq!(route.paths[0][3].node_features.le_flags(), &id_to_feature_flags(4));
+		assert_eq!(route.paths[0][3].channel_features.le_flags(), &id_to_feature_flags(11));
+
+		assert_eq!(route.paths[0][4].pubkey, nodes[6]);
+		assert_eq!(route.paths[0][4].short_channel_id, 8);
+		assert_eq!(route.paths[0][4].fee_msat, 100);
+		assert_eq!(route.paths[0][4].cltv_expiry_delta, 42);
+		assert_eq!(route.paths[0][4].node_features.le_flags(), &Vec::<u8>::new()); // We dont pass flags in from invoices yet
+		assert_eq!(route.paths[0][4].channel_features.le_flags(), &Vec::<u8>::new()); // We can't learn any flags from invoices, sadly
+	}
+
+	fn multi_hint_last_hops(nodes: &Vec<PublicKey>) -> Vec<RouteHint> {
+		let zero_fees = RoutingFees {
+			base_msat: 0,
+			proportional_millionths: 0,
+		};
+		vec![RouteHint(vec![RouteHintHop {
+			src_node_id: nodes[4].clone(),
+			short_channel_id: 11,
+			fees: zero_fees,
+			cltv_expiry_delta: (11 << 8) | 1,
+			htlc_minimum_msat: None,
+			htlc_maximum_msat: None,
+		}, RouteHintHop {
+			src_node_id: nodes[3].clone(),
+			short_channel_id: 8,
+			fees: zero_fees,
+			cltv_expiry_delta: (8 << 8) | 1,
+			htlc_minimum_msat: None,
+			htlc_maximum_msat: None,
```

Comment:
Is `last_hops_multi_hint_test` testing that a mutli-hop hint containing a public channel is used? And is `multi_hint_with_disabled_channels_last_hops` testing that a mutli-hop containing two private channels is used? If so, could you name the tests to reflect that? As of now, the second test seems to be named after how to trigger the behavior rather than after the behavior itself, IIUC.

Reply:
There are 3 tests:-
1. Checking empty hints are ignored - `ignores_empty_last_hops_test`.
2. Multi-hint last hops with all private channels - `multi_hint_last_hops_test`.
3. Multi-hint last hops test with one public channel used in hints - `last_hops_with_public_channel_test`.